File: Code/Csharp_Code/Chapter14_Lasers/LaserSimulator.cs
using System;
using System.Windows.Forms;
using System.Drawing;

public class LaserSimulator : Form
{
  private Label radiusLabel;
  private Label powerLabel;
  private Label absorptionLabel;

  private TextBox radiusTextBox;
  private TextBox powerTextBox;
  private TextBox absorptionTextBox;

  private Button startButton;
  private Button resetButton;

  private Panel drawingPanel;

  private double elapsedTime;
  private double explosionTime;
  private double airplaneX;
  private int laserX;
  private int laserZ;

  //  These fields are for the images used in the game.
  private Image airplaneIcon;
  private int airplaneIconWidth;
  private int airplaneIconHeight;
  private Image explosionIcon;
  private int explosionIconWidth;
  private int explosionIconHeight;

  //  The Timer is used to control the execution speed
  //  of the game.
  private Timer gameTimer;

  public LaserSimulator() {

    elapsedTime = 0.0;
    airplaneX = 0.0;
    laserX = 150;
    laserZ = 280;

    //  Set up some images and determine their dimensions
    airplaneIcon = Image.FromFile("airplaneCartoon.jpg");
    airplaneIconWidth = airplaneIcon.Width;
    airplaneIconHeight = airplaneIcon.Height;
    explosionIcon = Image.FromFile("explosionCartoon.jpg");
    explosionIconWidth = explosionIcon.Width;
    explosionIconHeight = explosionIcon.Height;

    //  Create a Timer object that will be used
    //  to slow the action down.
    gameTimer = new Timer();
    gameTimer.Interval = 100;  //  delay in milliseconds.
    gameTimer.Tick += new EventHandler(ActionPerformed);

    //  Create some Labels
    powerLabel = new Label();
    powerLabel.Text = "Beam power, J/cm^2";
    powerLabel.Font = new Font(powerLabel.Font, FontStyle.Bold);
    powerLabel.Top = 350;
    powerLabel.Left = 20;
    powerLabel.Width = 120;

    radiusLabel = new Label();
    radiusLabel.Text = "Beam radius, cm";
    radiusLabel.Font = new Font(radiusLabel.Font, FontStyle.Bold);
    radiusLabel.Top = 380;
    radiusLabel.Left = 20;
    radiusLabel.Width = 120;

    absorptionLabel = new Label();
    absorptionLabel.Text = "Absorption coefficient";
    absorptionLabel.Font = new Font(absorptionLabel.Font, FontStyle.Bold);
    absorptionLabel.Top = 410;
    absorptionLabel.Left = 20;
    absorptionLabel.Width = 120;

    //  Create TextBox objects to display the inputs.
    radiusTextBox = new TextBox();
    radiusTextBox.Width = 60;
    radiusTextBox.Text = "35.0";
    radiusTextBox.AutoSize = true;
    radiusTextBox.Top = radiusLabel.Top;
    radiusTextBox.Left = 150;

    powerTextBox = new TextBox();
    powerTextBox.Width = 60;
    powerTextBox.Text = "7.5e+6";
    powerTextBox.AutoSize = true;
    powerTextBox.Top = powerLabel.Top;
    powerTextBox.Left = 150;

    absorptionTextBox = new TextBox();
    absorptionTextBox.Width = 60;
    absorptionTextBox.Text = "0.75";
    absorptionTextBox.AutoSize = true;
    absorptionTextBox.Top = absorptionLabel.Top;
    absorptionTextBox.Left = 150;

    //  Create Button objects 
    int buttonHeight = 30;
    int buttonWidth = 50;
    int buttonLeft = 350;

    startButton = new Button();
    startButton.Text = "Start";
    startButton.Height = buttonHeight;
    startButton.Width = buttonWidth;
    startButton.Top = 370;
    startButton.Left = buttonLeft;
    startButton.Click += new EventHandler(StartButtonClicked);

    resetButton = new Button();
    resetButton.Text = "Reset";
    resetButton.Height = buttonHeight;
    resetButton.Width = buttonWidth;
    resetButton.Top = 410;
    resetButton.Left = buttonLeft;
    resetButton.Click += new EventHandler(ResetButtonClicked);

    //  Create a drawing panel.
    drawingPanel = new Panel();
    drawingPanel.Width = 351;
    drawingPanel.Height = 301;
    drawingPanel.Left = 50;
    drawingPanel.Top = 20;
    drawingPanel.BorderStyle = BorderStyle.FixedSingle;
    drawingPanel.MouseMove += new MouseEventHandler(MouseMoved);

    //  Add the GUI components to the Form
    this.Controls.Add(radiusLabel);
    this.Controls.Add(powerLabel);
    this.Controls.Add(absorptionLabel);
    this.Controls.Add(radiusTextBox);
    this.Controls.Add(powerTextBox);
    this.Controls.Add(absorptionTextBox);
    this.Controls.Add(startButton);
    this.Controls.Add(resetButton);
    this.Controls.Add(drawingPanel);

    // Set the size and title of the form
    this.Width = 450;
    this.Height = 500;
    this.Text = "Laser Simulator";

    //  Center the form on the screen and make
    //  it visible.
    this.StartPosition = FormStartPosition.CenterScreen;
    this.Visible = true;

    //  Update the GUI display
    UpdateDisplay();
  }

  //  Event handling method called when the mouse moves
  //  over the panel.
  public void MouseMoved(object source, MouseEventArgs evt) {

    laserX = evt.X;
    laserZ = evt.Y;
  }

  //  Event handling method for the "Reset" button
  public void ResetButtonClicked(object source, EventArgs e) {
    //  stop the timer.
    gameTimer.Stop();

    //  Reset values.
    elapsedTime = 0.0;
    airplaneX = 0.0;
    laserX = 150;
    laserZ = 280;

    //  Update the display.
    UpdateDisplay();
  }

  //  Event handling method for the "Start" button
  public void StartButtonClicked(object source, EventArgs e) {

    //  Get some initial quantities from the textfields.
    double power = Convert.ToDouble(powerTextBox.Text);
    double radius = Convert.ToDouble(radiusTextBox.Text);
    double absorption = Convert.ToDouble(absorptionTextBox.Text);

    //  Determine the beam area.
    double area = Math.PI*radius*radius;

    //  Compute beam contact time required to destroy airplane.
    explosionTime = 10000.0*area/(absorption*power);

    //  Update the display
    UpdateDisplay();

    //  Start the airplane using a Timer object
    gameTimer.Start();
  }

  //  This method redraws the GUI display.
  private void UpdateDisplay() {
    Graphics g = drawingPanel.CreateGraphics();
    int width = drawingPanel.Width - 1;
    int height = drawingPanel.Height - 1;

    //  Clear the current display.
    g.Clear(Color.White);

    //  Update the position of the soccerball on the screen.
    SolidBrush brush = new SolidBrush(Color.Black);
    Pen blackPen = new Pen(Color.Black, 1);

    //  Draw the ground
    g.DrawLine(blackPen, 0, height-10, width, height-10);

    //  Draw the airplane.
    g.DrawImage(airplaneIcon, (int)airplaneX, 20, 
                airplaneIconWidth, airplaneIconHeight);

    //  Draw the laser building.
    g.FillRectangle(brush, 140, height-30, 20, 20);

    //  Draw laser beam
    Pen redPen = new Pen(Color.Red, 2);
    g.DrawLine(redPen, 150, height-20, laserX, laserZ);

    //  If the time that the laser has been on the airplane is
    //  greater than the necessary time to destroy the airplane, 
    //  draw the explosion icon.
    if ( elapsedTime > explosionTime ) {
      g.DrawImage(explosionIcon, (int)airplaneX - 10, 10, 
                  explosionIconWidth, explosionIconHeight);      
    }

    //  Clean up the Graphics object.
    g.Dispose();
  }

  //  This method is called by the Timer every 0.1 seconds.
  public void ActionPerformed(object source, EventArgs e) {

    //  set the time increment. 
    double timeIncrement = 0.1;

    //  If the laser is hitting the airplane increment
    //  the elapsedTime variable. If not, set the elapsedTime
    //  variable to zero.
    if ( laserX > airplaneX && laserX < airplaneX + airplaneIconWidth &&
         laserZ > 10 && laserZ < 10 + airplaneIconHeight ) {
      elapsedTime = elapsedTime + timeIncrement;
    }
    else {
      elapsedTime = 0.0;
    }

    //  The airplane cartoon moves across the screen at a 
    //  constant velocity of 20 pixels/second.
    airplaneX = airplaneX + 20.0*timeIncrement;

    //  Update the display
    UpdateDisplay();

    //  If the time that the laser has been on the airplane is
    //  greater than the necessary time to destroy the airplane, 
    //  stop the simulation.
    if ( elapsedTime > explosionTime ) {
      gameTimer.Stop();      
    }
  }

  static void Main() {
    Application.Run(new LaserSimulator());
  }
}

File: Code/Java_Code/Chapter14_Lasers/LaserSimulator.java
import javax.swing.*;
import java.awt.*;
import javax.swing.border.BevelBorder;
import java.awt.event.*;
import javax.swing.Timer;

public class LaserSimulator extends JFrame implements ActionListener, 
               MouseListener, MouseMotionListener
{
  private JTextField radiusTextField;
  private JTextField powerTextField;
  private JTextField absorptionTextField;

  private JLabel radiusLabel;
  private JLabel powerLabel;
  private JLabel absorptionLabel;

  private JButton startButton;
  private JButton resetButton;
  private JPanel drawingPanel;
  private GridBagConstraints gbc;

  private double elapsedTime;
  private double explosionTime;
  private double airplaneX;
  private int laserX;
  private int laserZ;

  //  Laser beam fields.
  double power;
  double radius;
  double absorption;
  double area;

  //  These fields are for the images used in the game.
  private ImageIcon airplaneIcon;
  private int airplaneIconWidth;
  private int airplaneIconHeight;
  private ImageIcon explosionIcon;
  private int explosionIconWidth;
  private int explosionIconHeight;

  //  These elements are used to control the execution
  //  speed of the game. Without them, the game would
  //  run too quickly.
  private GameUpdater gameUpdater;
  private Timer gameTimer;

  public LaserSimulator() {

    elapsedTime = 0.0;
    airplaneX = 0.0;
    laserX = 150;
    laserZ = 280;

    //  Create a Timer object that will be used
    //  to slow the action down and an ActionListener
    //  that the Timer will call. 
    gameUpdater = new GameUpdater();
    gameTimer = new Timer(100, gameUpdater);

    //  Set up some images and determine their dimensions
    airplaneIcon = new ImageIcon("airplaneCartoon.jpg");
    airplaneIconWidth = airplaneIcon.getIconWidth();
    airplaneIconHeight = airplaneIcon.getIconHeight();
    explosionIcon = new ImageIcon("explosionCartoon.jpg");
    explosionIconWidth = explosionIcon.getIconWidth();
    explosionIconHeight = explosionIcon.getIconHeight();

    //  Create JTextField objects to input the initial
    //  conditions.
    radiusTextField = new JTextField("35.0",6);
    powerTextField = new JTextField("7.5e+6",6);
    absorptionTextField = new JTextField("0.75",6);

    //  Create some JLabels
    radiusLabel = new JLabel("Beam radius, cm");
    powerLabel = new JLabel("Beam power, J/cm^2");
    absorptionLabel = new JLabel("Absorption coefficient");

    //  Create a JButton that will start the ball moving
    startButton = new JButton("Start");
    startButton.setBorder(new BevelBorder(BevelBorder.RAISED));
    startButton.setPreferredSize(new Dimension(60,35));
    startButton.addActionListener(this);

    //  Create a JButton that will update the drawing area.
    resetButton = new JButton("Reset");
    resetButton.setBorder(new BevelBorder(BevelBorder.RAISED));
    resetButton.setPreferredSize(new Dimension(60,35));
    resetButton.addActionListener( new ActionListener() {
      public void actionPerformed(ActionEvent ae) {
        //  stop the timer and reset the time.
        gameTimer.stop();

        //  Reset values.
        elapsedTime = 0.0;
        airplaneX = 0.0;
        laserX = 150;
        laserZ = 280;

        //  Update the display.
        updateDisplay();
      }  
    });

    //  Create a JPanel that will display the results
    drawingPanel = new JPanel();
    drawingPanel.setPreferredSize(new Dimension(351, 301));
    drawingPanel.addMouseMotionListener(this);
    drawingPanel.addMouseListener(this);

    //  Place components on a panel using a GridBagLayout
    JPanel westPanel = new JPanel();
    GridBagLayout gridBagLayout1 = new GridBagLayout();
    westPanel.setLayout(gridBagLayout1);

    int col;
    int row;
    int numCol = 1;
    int numRow = 1;
    Insets insets = new Insets(5, 3, 5, 3);

    row = 0;
    col = 0;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.EAST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(powerLabel, gbc);

    col = 1;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.WEST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(powerTextField, gbc);

    row = 1;
    col = 0;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.EAST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(radiusLabel, gbc);

    col = 1;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.WEST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(radiusTextField, gbc);

    row = 2;
    col = 0;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.EAST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(absorptionLabel, gbc);

    col = 1;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.WEST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(absorptionTextField, gbc);

    row = 3;
    col = 0;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.EAST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(startButton, gbc);

    row = 4;
    col = 0;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.EAST,
                 GridBagConstraints.NONE, insets, 0, 0);
    gridBagLayout1.setConstraints(resetButton, gbc);

    westPanel.add(powerLabel);
    westPanel.add(powerTextField);
    westPanel.add(radiusLabel);
    westPanel.add(radiusTextField);
    westPanel.add(absorptionLabel);
    westPanel.add(absorptionTextField);
    westPanel.add(startButton);
    westPanel.add(resetButton);

    //  The drawing panel.
    JPanel northPanel = new JPanel();
    GridBagLayout gridBagLayout3 = new GridBagLayout();
    northPanel.setLayout(gridBagLayout3);

    row = 0;
    col = 0;
    gbc = new GridBagConstraints(col, row, numCol, numRow,
                 0.0, 0.0, GridBagConstraints.CENTER,
                 GridBagConstraints.NONE, 
                 new Insets(10, 10, 10, 20), 0, 0);
    gridBagLayout3.setConstraints(drawingPanel, gbc);

    northPanel.add(drawingPanel);

    //  Add the JPanel objects to the content pane
    getContentPane().setLayout(new BorderLayout());
    getContentPane().add(westPanel, BorderLayout.WEST);
    getContentPane().add(northPanel, BorderLayout.NORTH);

    //  Add a title to the JFrame, size it, and make it visible.
    setTitle("Laser Simulator");
    setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
    setBounds(100,100,450,550);
    setVisible(true);

    //  Update the GUI display
    updateDisplay(); 
  }

  //  The actionPerformed() method is called when 
  //  the "Start" button is pressed. 
  public void actionPerformed(ActionEvent event) {

    //  Get some initial quantities from the textfields.
    double power = Double.parseDouble(powerTextField.getText());
    double radius = Double.parseDouble(radiusTextField.getText());
    double absorption = Double.parseDouble(absorptionTextField.getText());

    //  Determine the beam area.
    double area = Math.PI*radius*radius;

    //  Compute beam contact time required to destroy airplane.
    explosionTime = 10000.0*area/(absorption*power);

    //  Start the airplane using a Timer object
    gameTimer.start();
  }

  //  These methods are used to update the position of the
  //  laser.
  public void mouseDragged(MouseEvent me) {
    laserX = me.getX();
    laserZ = me.getY();
  }

  public void mouseClicked(MouseEvent me) {
    laserX = me.getX();
    laserZ = me.getY();
  }

  //  These methods have to be included as part of the 
  //  listener interfaces even though they aren't used.
  public void mouseMoved(MouseEvent me) {}
  public void mouseEntered(MouseEvent me) {}
  public void mouseExited(MouseEvent me) {}
  public void mouseReleased(MouseEvent me) {}
  public void mousePressed(MouseEvent me) {}

  //  This method redraws the GUI display.
  private void updateDisplay() {
    Graphics g = drawingPanel.getGraphics();
    int width = drawingPanel.getWidth() - 1;
    int height = drawingPanel.getHeight() - 1;

    g.clearRect(0, 0, width, height);
    g.setColor(Color.WHITE);
    g.fillRect(0, 0, width, height); 

    g.setColor(Color.BLACK);
    g.drawLine(0, height-10, width, height-10);

    //  Draw the airplane.
    g.drawImage(airplaneIcon.getImage(), (int)airplaneX, 20, 
                airplaneIconWidth, airplaneIconHeight, drawingPanel);

    //  Draw the laser building.
    g.fillRect(140, height-30, 20, 20);

    //  Draw laser beam
    g.setColor(Color.RED);
    g.drawLine(150, height-20, laserX, laserZ);

    //  If the time that the laser has been on the airplane is
    //  greater than the necessary time to destroy the airplane, 
    //  draw the explosion icon.
    if ( elapsedTime > explosionTime ) {
      g.setColor(Color.BLACK);
      g.drawImage(explosionIcon.getImage(), (int)airplaneX - 10, 10, 
                explosionIconWidth, explosionIconHeight, drawingPanel);      
    }

  }

  public static void main(String args[]) {
    LaserSimulator gui = new LaserSimulator();
  }

  //  This ActionListener is called by the Timer
  class GameUpdater implements ActionListener {
    public void actionPerformed(ActionEvent event) {

      //  set the time increment. 
      double timeIncrement = 0.1;

      //  If the laser is hitting the airplane increment
      //  the elapsedTime variable. If not, set the elapsedTime
      //  variable to zero.
      if ( laserX > airplaneX && laserX < airplaneX + airplaneIconWidth &&
           laserZ > 10 && laserZ < 10 + airplaneIconHeight ) {
        elapsedTime = elapsedTime + timeIncrement;
      }
      else {
        elapsedTime = 0.0;
      }

      //  The airplane cartoon moves across the screen at a 
      //  constant velocity of 20 pixels/second.
      airplaneX = airplaneX + 20.0*timeIncrement;

      //  Update the display
      updateDisplay();

      //  If the time that the laser has been on the airplane is
      //  greater than the necessary time to destroy the airplane, 
      //  stop the simulation.
      if ( elapsedTime > explosionTime ) {
        gameTimer.stop();      
      }

    }
  }
}